R/conditional-transform.R

```diff
@@ -0,0 +1,150 @@
+#'                                        name = "Opinoin of Cats")
+#' }
+#' @export
+conditionalTransform <- function (..., data, else_condition=NA, type="categorical", categories=NULL) {
+    dots <- list(...)
+    is_formula <- function (x) inherits(x, "formula")
+    formulas <- Filter(is_formula, dots) 
+    var_def <- Filter(Negate(is_formula), dots)
+    n <- length(formulas)
+    
+    if (n == 0) {
+        halt("no conditions have been supplied; please supply formulas as conditions.")
+    }
+    if (!type %in% c("categorical", "text", "numeric")){
+        halt("type must be either ", dQuote("categorical"), ", ", dQuote("text"), ", or ", dQuote("numeric"))
+    }
+    if (type != "categorical" & !is.null(categories)){
+        halt("type is not ", dQuote("categorical"), " ignoring ", dQuote("categories"))
+    }
+    
+    cases <- vector("list", n)
+    values <- vector("list", n)
+    for (i in seq_len(n)) {
+        formula <- formulas[[i]]
+        if (!inherits(formula, "formula") || length(formula) != 3) {
+            halt("The case provided is not a formula: ", deparseAndFlatten(formula))
+        }
+        
+        cases[[i]] <- evalLHS(formula, data)
+        if (!inherits(cases[[i]], "CrunchLogicalExpr")) {
+            halt("The LHS provided is not a CrunchLogicalExpr: ", formulaLHS(formula))
+        }
+        values[[i]] <- evalRHS(formula, data)
+    }
+    case_indices <- lapply(cases, which)
+    
+    # dedpulicate indices, favoring the first true condition
+    case_indices <- lapply(seq_along(case_indices), function(i) {
+        setdiff(case_indices[[i]], unlist(case_indices[seq_len(i-1)]))
+    })
+    
+    # grab the values needed from source variables
+    values_to_fill <- Map(function(ind, var) {
+        if (inherits(var, c("CrunchVariable", "CrunchExpr"))) {
+            # grab the variable contents at inds
+            return(as.vector(var[ind]))
+        } else {
+            # if var isn't a crunch variable or expression, just return var
+            return(var)
+        }
+    }, ind = case_indices, var = values)
+    
+    # setup NAs for as default
+    n_rows <- nrow(CrunchDataset(crGET(datasetReference(cases[[1]]))))
+    result <- rep(else_condition, n_rows)
+    
+    # fill values
+    for (i in seq_along(case_indices)) {
+        if (type == "numeric") {
+            trans <- as.numeric
+        } else {
+            trans <- as.character
+        }
+        vals <- trans(values_to_fill[[i]])
+        result[case_indices[[i]]] <- vals
+    }
+ 
+    if (type == "categorical") {
+        # if categories are supplied and there are any
+        if (missing(categories)) {
+            result <- factor(result)
+            # if categories aren't a Categories object,
+            # make categories from names
+            categories <- Categories(data = categoriesFromLevels(levels(result)))
+            categories <- c(categories, Category(data=.no.data))
```

Comment:
I'd factor out this "No data" category adding since you're doing it in every case. And look to how you can encapsulate this logic better in the few other existing parts of the code where this behavior happens (search for `.no.data`).

Reply:
Absolutely, I've added an `ensureNoDataCategory()` function that adds the No Data category if there's not one already. I've put it in the few places it would fit elsewhere in the package as well.